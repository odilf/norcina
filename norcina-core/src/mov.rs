use std::fmt;

use enum_dispatch::enum_dispatch;

#[enum_dispatch]
pub trait Move {}

#[enum_dispatch]
pub trait InvertibleMove: Move {
    /// The move that undoes `self`.
    fn inverse(&self) -> Self;
}

#[enum_dispatch]
pub trait RandomMove: Move {
    /// Generates a random move.
    fn random(rng: &mut impl rand::Rng) -> Self;
}

#[enum_dispatch]
pub trait DisplayMove: Move + fmt::Display {}
impl<M: Move + fmt::Display> DisplayMove for M {}
