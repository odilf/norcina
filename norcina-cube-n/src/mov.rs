use std::{fmt, ops};

use crate::math::{Axis, Direction, Face};

#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Amount {
    Single = 1,
    Double = 2,
    Reverse = 3,
}

impl Amount {
    #[inline]
    pub const fn u8(self) -> u8 {
        self as u8
    }

    #[inline]
    pub const fn from_u8(x: u8) -> Self {
        // TODO: Transmute and stuff
        match x {
            1 => Self::Single,
            2 => Self::Double,
            3 => Self::Reverse,
            _ => panic!("Invalid amount."),
        }
    }

    /// The amount that does the opposite of itself.
    pub const fn reverse(self) -> Amount {
        // TODO: This can be done more efficiently with bit-twiddling, but is it worth it...
        match self {
            Amount::Single => Amount::Reverse,
            Amount::Double => Amount::Double,
            Amount::Reverse => Amount::Single,
        }
    }

    pub fn iter() -> impl Iterator<Item = Amount> {
        [Amount::Single, Amount::Double, Amount::Reverse].into_iter()
    }
}

impl ops::Mul<Direction> for Amount {
    type Output = Amount;
    fn mul(self, rhs: Direction) -> Self::Output {
        // TODO: Is it possible to remove this branch?
        match rhs {
            Direction::Positive => self,
            #[allow(clippy::suspicious_arithmetic_impl)]
            Direction::Negative => Self::from_u8(4 - self.u8()),
        }
    }
}

#[repr(transparent)]
#[derive(Clone, Copy, PartialEq, Eq)]
pub struct Move {
    /// Packed field: `---aafff`
    data: u8,
}

impl Move {
    pub const fn new(face: Face, amount: Amount) -> Move {
        Self {
            data: face.u8() + (amount.u8() << 3),
        }
    }

    pub const fn face(self) -> Face {
        Face::from_u8(self.data & 0b111)
    }

    pub const fn amount(self) -> Amount {
        Amount::from_u8(self.data >> 3)
    }

    pub const fn axis(self) -> Axis {
        self.face().axis()
    }

    /// Enumerates all possible moves.
    pub fn iter() -> impl Iterator<Item = Self> {
        Face::iter().flat_map(|face| Amount::iter().map(move |amount| Move::new(face, amount)))
    }

    pub const ALL: [Move; 18] = {
        use moves::*;
        [
            R, R2, RP, U, U2, UP, F, F2, FP, L, L2, LP, D, D2, DP, B, B2, BP,
        ]
    };
}

impl fmt::Debug for Move {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Move")
            .field("face", &self.face())
            .field("amount", &self.amount())
            .finish()
    }
}

impl fmt::Display for Move {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let amount_str = match self.amount() {
            Amount::Single => " ",
            Amount::Double => "2",
            Amount::Reverse => "'",
        };

        write!(f, "{}{}", self.face(), amount_str)
    }
}

impl norcina_core::Move for Move {}
impl norcina_core::mov::InvertibleMove for Move {
    fn inverse(&self) -> Self {
        // TODO: This can be done more efficiently with bit-twiddling.
        Self::new(self.face(), self.amount().reverse())
    }
}

pub mod moves {
    macro_rules! generate_moves {
        ([$($face:tt $amount:tt $name:ident),*]) => {
            $(
                pub const $name: Move = Move::new(
                    generate_moves!(@face $face),
                    generate_moves!(@amount $amount)
                );
            )*
        };

        (@amount 1) => { Amount::Single };
        (@amount 2) => { Amount::Double };
        (@amount 3) => { Amount::Reverse };

        (@face R) => { Face::R };
        (@face U) => { Face::U };
        (@face F) => { Face::F };
        (@face L) => { Face::L };
        (@face D) => { Face::D };
        (@face B) => { Face::B };
    }

    use super::Amount;
    use super::Face;
    use super::Move;

    generate_moves!([
        R 1 R, R 2 R2, R 3 RP,
        U 1 U, U 2 U2, U 3 UP,
        F 1 F, F 2 F2, F 3 FP,
        L 1 L, L 2 L2, L 3 LP,
        D 1 D, D 2 D2, D 3 DP,
        B 1 B, B 2 B2, B 3 BP
    ]);
}

#[macro_export]
macro_rules! alg {
    (@ $mov:tt) => { $mov };

    ($($mov:tt)*) => {{
        use $crate::mov::moves::*;
        [$(alg!(@ $mov)),*]
    }};
}

#[cfg(feature = "quickcheck")]
mod quickcheck_impl {
    use super::*;

    use quickcheck::{Arbitrary, Gen};

    impl Arbitrary for Amount {
        fn arbitrary(g: &mut Gen) -> Self {
            *g.choose(&[Amount::Single, Amount::Double, Amount::Reverse])
                .unwrap()
        }
    }

    impl Arbitrary for Move {
        fn arbitrary(g: &mut Gen) -> Self {
            let face = Face::arbitrary(g);
            let amount = Amount::arbitrary(g);
            Move::new(face, amount)
        }
    }
}
