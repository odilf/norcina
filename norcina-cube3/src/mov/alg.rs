pub mod algs {
    macro_rules! declare_alg {
        ($name:tt = $($alg:tt)*) => {
            declare_alg!(@inner
                (name=$name)
                (length=0)
                (acc=[])
                (rest=[$($alg),*])
            );
        };

        (@inner (name=$name:ident) (length=$length:expr) (acc=[$($alg:tt),*]) (rest=[ ])) => {
            use $crate::mov::moves::*;
            pub const $name: [Move; $length] = [$($alg),*];
        };

        (@inner
            (name=$name:tt)
            (length=$length:expr)
            (acc=[$($acc:tt)*])
            (rest=[$alg_head:tt $(, $($alg_tail:tt)+)? ])
        ) => {
            declare_alg!(@inner
                (name=$name)
                (length=$length + 1)
                (acc=[$($acc)* $alg_head])
                (rest=[ $($($alg_tail),+)? ])
            );
        };

    }

    use norcina_cube_n::alg;

    use crate::Move;

    // declare_alg!(SIMPLE = R);
    // declare_alg!(@inner (name=SEXY) (length=0) (acc=[]) (rest=[R, U]));
    // declare_alg!(@inner (name=SEXY) (length=0 + 1) (acc=[R]) (rest=[U]));
    // declare_alg!(@inner (name=SEXY) (length=0 + 1 + 1) (acc=[R, U]) (rest=[ ]));

    // declare_alg!(SEXY = R U);
    // declare_alg!(SEXY = R U RP UP);
    // pub const SEXY: [Move; 4] = alg![R U RP UP];
    pub const SLEDGEHAMMER: [Move; 4] = alg!(RP F R FP);

    pub mod oll {
        use super::*;
        pub const T: [Move; 14] = alg!(R U RP UP RP F R2 UP RP UP R U RP FP);
        pub const J: [Move; 13] = alg!(R U RP F R U RP UP RP FP R2 UP RP);
        pub const U: [Move; 11] = U_A;
        pub const U_A: [Move; 11] = alg!(R2 UP RP UP R U R U R UP R);
        pub const U_B: [Move; 11] = alg!(RP U RP UP RP UP RP U R U R2);
    }

    pub const CHECKER: [Move; 6] = alg!(R2 L2 U2 D2 F2 B2);

    // TODO: Concat or extend algs
    // pub const J_AUF: [Move; 14] = [J, alg!(UP)].concat();
}
