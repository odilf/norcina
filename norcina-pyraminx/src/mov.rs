use std::{fmt, mem};

use crate::piece::Vertex;

// TODO: Maybe make a struct with a u8...
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Amount {
    Single = 0,
    Reverse = 1,
}

impl Amount {
    #[inline(always)]
    pub const fn u8(self) -> u8 {
        self as u8
    }

    /// # Safety
    ///
    /// `data` must be either 0 or 1.
    pub const unsafe fn from_u8_unchecked(data: u8) -> Self {
        debug_assert!(data < 2);
        // SAFETY: Precodnition is that it is either 0 or 1.
        unsafe { mem::transmute(data) }
    }

    pub const fn from_u8(data: u8) -> Self {
        assert!(data < 2);
        // SAFETY: We just checked that `data` is less than 2, so it is either 0 or 1.
        unsafe { mem::transmute(data) }
    }

    pub const fn from_u8_mod2(data: u8) -> Amount {
        // SAFETY: We take modulo 2.
        unsafe { Self::from_u8_unchecked(data & 0b1) }
    }
}

#[repr(transparent)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct CoreMove {
    // Packed field: -----avv
    data: u8,
}

impl CoreMove {
    pub const fn new(vertex: Vertex, amount: Amount) -> Self {
        Self {
            data: vertex.u8() + (amount.u8() << 2),
        }
    }

    pub const ALL: [CoreMove; 8] = [
        CoreMove::new(Vertex::R, Amount::Single),
        CoreMove::new(Vertex::U, Amount::Single),
        CoreMove::new(Vertex::L, Amount::Single),
        CoreMove::new(Vertex::B, Amount::Single),
        CoreMove::new(Vertex::R, Amount::Reverse),
        CoreMove::new(Vertex::U, Amount::Reverse),
        CoreMove::new(Vertex::L, Amount::Reverse),
        CoreMove::new(Vertex::B, Amount::Reverse),
    ];

    #[inline(always)]
    pub const fn vertex(self) -> Vertex {
        Vertex::from_u8_mod4(self.data & 0b011)
    }

    #[inline(always)]
    pub const fn amount(self) -> Amount {
        Amount::from_u8_mod2(self.data >> 2)
    }
}

#[repr(transparent)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Move {
    // Packed field: ----tavv
    data: u8,
}

impl Move {
    pub const R: Self = Self::new(Vertex::R, Amount::Single, false);
    pub const RP: Self = Self::new(Vertex::R, Amount::Reverse, false);
    pub const L: Self = Self::new(Vertex::L, Amount::Single, false);
    pub const LP: Self = Self::new(Vertex::L, Amount::Reverse, false);
    pub const B: Self = Self::new(Vertex::B, Amount::Single, false);
    pub const BP: Self = Self::new(Vertex::B, Amount::Reverse, false);
    pub const U: Self = Self::new(Vertex::U, Amount::Single, false);
    pub const UP: Self = Self::new(Vertex::U, Amount::Reverse, false);

    pub const fn new(vertex: Vertex, amount: Amount, tip: bool) -> Self {
        Self {
            data: vertex.u8() + (amount.u8() << 2) + ((tip as u8) << 3),
        }
    }
    pub const fn is_tip_move(self) -> bool {
        (self.data & 0b1000) != 0
    }

    pub const fn core(self) -> CoreMove {
        CoreMove {
            data: self.data & 0b0111,
        }
    }

    pub const fn toggle_tip(self) -> Self {
        Self {
            data: self.data ^ 0b1000,
        }
    }
}

impl fmt::Display for Move {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let addendum = if self.core().amount() == Amount::Single {
            ' '
        } else {
            '\''
        };

        let main = match (self.core().vertex(), self.is_tip_move()) {
            (Vertex::U, false) => "U",
            (Vertex::U, true) => "u",
            (Vertex::L, false) => "L",
            (Vertex::L, true) => "l",
            (Vertex::B, false) => "B",
            (Vertex::B, true) => "b",
            (Vertex::R, false) => "R",
            (Vertex::R, true) => "r",
            (_, _) => unreachable!(),
        };

        write!(f, "{main}{addendum}")
    }
}
