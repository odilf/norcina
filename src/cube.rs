use std::fmt::{self, Write as _};

use owo_colors::{OwoColorize, Rgb};

use crate::{
    corner::{self, Corner, CornerPosition},
    edge::{self, Edge, EdgePosition},
    math::Face,
    mov::Move,
};

#[derive(Clone, Copy, PartialEq, Eq)]
pub struct Cube {
    corners: [Corner; 8],
    edges: [Edge; 12],
}

impl Cube {
    pub const SOLVED: Self = Cube {
        corners: [
            Corner::solved(0),
            Corner::solved(1),
            Corner::solved(2),
            Corner::solved(3),
            Corner::solved(4),
            Corner::solved(5),
            Corner::solved(6),
            Corner::solved(7),
        ],
        edges: [
            Edge::solved(0),
            Edge::solved(1),
            Edge::solved(2),
            Edge::solved(3),
            Edge::solved(4),
            Edge::solved(5),
            Edge::solved(6),
            Edge::solved(7),
            Edge::solved(8),
            Edge::solved(9),
            Edge::solved(10),
            Edge::solved(11),
        ],
    };

    fn sticker_at(self, face: Face, up: Face, col: i32, row: i32) -> Sticker {
        // Center sticker
        if col == 1 && row == 1 {
            return face;
        }

        let side = up.cross(face);

        if (col + row) % 2 == 0 {
            let faces = [
                face,
                if row == 0 { up } else { up.opposite() },
                if col == 0 { side.opposite() } else { side },
            ];

            let position = CornerPosition::from_faces(faces);
            let piece = position.pick(self.corners);
            corner::sticker(piece, position, face)
        } else {
            let other_face = match (row, col) {
                (0, 1) => up,
                (1, 0) => side.opposite(),
                (1, 2) => side,
                (2, 1) => up.opposite(),
                _ => unreachable!(),
            };

            let position = EdgePosition::from_faces([face, other_face]);
            let piece = position.pick(&self.edges);
            edge::sticker(piece, position, face)
        }
    }

    pub fn mov(self, mov: Move) -> Self {
        Self {
            corners: corner::move_pieces(self.corners, mov),
            edges: edge::move_pieces(self.edges, mov),
        }
    }
}

pub type Sticker = Face;
pub type ColorScheme = fn(Face) -> Rgb;

const DEFAULT_COLOR_SCHEME: ColorScheme = |face| match face {
    Face::R => Rgb(217, 39, 39),
    Face::U => Rgb(250, 250, 250),
    Face::F => Rgb(109, 242, 116),
    Face::L => Rgb(255, 153, 12),
    Face::D => Rgb(255, 224, 0),
    Face::B => Rgb(79, 123, 212),
};

impl fmt::Debug for Cube {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        struct CornersDebug([Corner; 8]);
        impl fmt::Debug for CornersDebug {
            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                let mut dl = f.debug_list();
                for (i, corner) in self.0.into_iter().enumerate() {
                    let position = CornerPosition::from_index(i as u8);
                    dl.entry(&format_args!("{corner} is at {position}"));
                }

                dl.finish()
            }
        }

        struct EdgesDebug([Edge; 12]);
        impl fmt::Debug for EdgesDebug {
            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                let mut dl = f.debug_list();
                for (i, edge) in self.0.into_iter().enumerate() {
                    let position = EdgePosition::from_index(i as u8);
                    dl.entry(&format_args!("{edge} is at {position}"));
                }

                dl.finish()
            }
        }

        f.debug_struct("Cube")
            .field("corners", &CornersDebug(self.corners))
            .field("edges", &EdgesDebug(self.edges))
            .finish()
    }
}

impl fmt::Display for Cube {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let write = |f: &mut fmt::Formatter<'_>, sticker| {
            write!(f, "{}", "██".color((DEFAULT_COLOR_SCHEME)(sticker)))
        };

        let pad = |f: &mut fmt::Formatter<'_>| f.write_str("      ");

        for u_row in 0..3 {
            pad(f)?;
            for u_col in 0..3 {
                write(f, self.sticker_at(Face::U, Face::B, u_col, u_row))?;
            }
            f.write_char('\n')?;
        }

        for lfr_row in 0..3 {
            for l_col in 0..3 {
                write(f, self.sticker_at(Face::L, Face::U, l_col, lfr_row))?;
            }
            for f_col in 0..3 {
                write(f, self.sticker_at(Face::F, Face::U, f_col, lfr_row))?;
            }
            for r_col in 0..3 {
                write(f, self.sticker_at(Face::R, Face::U, r_col, lfr_row))?;
            }

            f.write_char('\n')?;
        }

        for d_row in 0..3 {
            pad(f)?;
            for d_col in 0..3 {
                write(f, self.sticker_at(Face::D, Face::F, d_col, d_row))?;
            }
            f.write_char('\n')?;
        }
        for b_row in 0..3 {
            pad(f)?;
            for b_col in 0..3 {
                write(f, self.sticker_at(Face::B, Face::D, b_col, b_row))?;
            }
            f.write_char('\n')?;
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use quickcheck::{quickcheck, Arbitrary, Gen};

    use super::*;

    impl Arbitrary for Cube {
        fn arbitrary(g: &mut Gen) -> Self {
            Self {
                corners: [Corner::arbitrary(g); 8],
                edges: [Edge::arbitrary(g); 12],
            }
        }
    }
}
