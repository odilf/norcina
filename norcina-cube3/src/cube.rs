use std::fmt::{self, Write as _};

use norcina_cube_n::math::Face;
use norcina_cube_n::mov::Move;
use norcina_cube_n::piece::{
    corner::{self, Corner, CornerPosition},
    edge::{self, Edge, EdgePosition},
};
use owo_colors::{OwoColorize, Rgb};

#[derive(Clone, Copy, PartialEq, Eq, Hash)]
pub struct Cube {
    // TODO: Is it a problem if these are pub? It might be...
    pub corners: [Corner; 8],
    pub edges: [Edge; 12],
}

impl Cube {
    pub const SOLVED: Self = Cube {
        corners: Corner::SOLVED,
        edges: Edge::SOLVED,
    };

    pub fn random_with_rng(rng: &mut impl rand::Rng) -> Self {
        let mut corners = Corner::random(rng);
        let mut edges = Edge::random(rng);

        let corner_swap_parity = Corner::count_swaps(corners) % 2;
        let edge_swap_parity = Edge::count_swaps(edges) % 2;

        dbg!(corner_swap_parity, edge_swap_parity);

        if corner_swap_parity != edge_swap_parity {
            // TODO: Could we always swap the same two arbtirary pieces, or
            // would that stop it from being uniform distribution?
            if rng.random() {
                let i = rng.random_range(0..8);
                let mut j = rng.random_range(0..7);
                if j >= i {
                    j += 1;
                }

                corners.swap(i, j);
            } else {
                let i = rng.random_range(0..12);
                let mut j = rng.random_range(0..11);
                if j >= i {
                    j += 1;
                }

                edges.swap(i, j);
            }
        }

        Cube { corners, edges }
    }

    pub fn random() -> Self {
        Self::random_with_rng(&mut rand::rng())
    }

    fn sticker_at(self, face: Face, up: Face, col: i32, row: i32) -> Sticker {
        // Center sticker
        if col == 1 && row == 1 {
            return face;
        }

        let side = up.cross(face);

        if (col + row) % 2 == 0 {
            let faces = [
                face,
                if row == 0 { up } else { up.opposite() },
                if col == 0 { side.opposite() } else { side },
            ];

            let position = CornerPosition::from_faces(faces);
            let piece = position.pick(self.corners);
            corner::sticker(piece, position, face)
        } else {
            let other_face = match (row, col) {
                (0, 1) => up,
                (1, 0) => side.opposite(),
                (1, 2) => side,
                (2, 1) => up.opposite(),
                _ => unreachable!(),
            };

            let position = EdgePosition::from_faces([face, other_face]);
            let piece = position.pick(&self.edges);
            edge::sticker(piece, position, face)
        }
    }

    pub fn mov_single(self, mov: Move) -> Self {
        Self {
            corners: corner::move_pieces(self.corners, mov),
            edges: edge::move_pieces(self.edges, mov),
        }
    }

    pub fn mov(mut self, alg: impl IntoIterator<Item = Move>) -> Self {
        for mov in alg {
            self = self.mov_single(mov);
        }
        self
    }

    pub const fn is_solved(self) -> bool {
        matches!(self, Self::SOLVED)
    }

    /// An iterator of tuples of every state that is reachable from this state, and the number of moves it takes to reach it.
    pub fn neighbors(self) -> impl Iterator<Item = (Move, Self)> {
        Move::iter().map(move |mov| (mov, self.mov_single(mov)))
    }

    pub fn corners(self) -> impl Iterator<Item = (CornerPosition, Corner)> {
        self.corners.into_iter().enumerate().map(|(i, state)| {
            (
                unsafe { CornerPosition::from_index_unchecked(i as u8) },
                state,
            )
        })
    }

    pub fn edges(self) -> impl Iterator<Item = (EdgePosition, Edge)> {
        self.edges.into_iter().enumerate().map(|(i, state)| {
            (
                unsafe { EdgePosition::from_index_unchecked(i as u8) },
                state,
            )
        })
    }
}

pub type Sticker = Face;
pub type ColorScheme = fn(Face) -> Rgb;

const DEFAULT_COLOR_SCHEME: ColorScheme = |face| match face {
    Face::R => Rgb(217, 39, 39),
    Face::U => Rgb(250, 250, 250),
    Face::F => Rgb(109, 242, 116),
    Face::L => Rgb(255, 153, 12),
    Face::D => Rgb(255, 224, 0),
    Face::B => Rgb(79, 123, 212),
};

impl fmt::Debug for Cube {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        struct CornersDebug([Corner; 8]);
        impl fmt::Debug for CornersDebug {
            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                let mut dl = f.debug_list();
                for (i, corner) in self.0.into_iter().enumerate() {
                    let position = CornerPosition::from_index(i as u8);
                    dl.entry(&format_args!("{corner} is at {position}"));
                }

                dl.finish()
            }
        }

        struct EdgesDebug([Edge; 12]);
        impl fmt::Debug for EdgesDebug {
            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                let mut dl = f.debug_list();
                for (i, edge) in self.0.into_iter().enumerate() {
                    let position = EdgePosition::from_index(i as u8);
                    dl.entry(&format_args!("{edge} is at {position}"));
                }

                dl.finish()
            }
        }

        f.debug_struct("Cube")
            .field("corners", &CornersDebug(self.corners))
            .field("edges", &EdgesDebug(self.edges))
            .finish()
    }
}

impl fmt::Display for Cube {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let write = |f: &mut fmt::Formatter<'_>, sticker| {
            write!(f, "{}", "██".color((DEFAULT_COLOR_SCHEME)(sticker)))
        };

        let pad = |f: &mut fmt::Formatter<'_>| f.write_str("      ");

        for b_row in 0..3 {
            pad(f)?;
            for b_col in 0..3 {
                write(f, self.sticker_at(Face::B, Face::D, b_col, b_row))?;
            }
            f.write_char('\n')?;
        }
        for u_row in 0..3 {
            pad(f)?;
            for u_col in 0..3 {
                write(f, self.sticker_at(Face::U, Face::B, u_col, u_row))?;
            }
            f.write_char('\n')?;
        }

        for lfr_row in 0..3 {
            for l_col in 0..3 {
                write(f, self.sticker_at(Face::L, Face::U, l_col, lfr_row))?;
            }
            for f_col in 0..3 {
                write(f, self.sticker_at(Face::F, Face::U, f_col, lfr_row))?;
            }
            for r_col in 0..3 {
                write(f, self.sticker_at(Face::R, Face::U, r_col, lfr_row))?;
            }

            f.write_char('\n')?;
        }

        for d_row in 0..3 {
            pad(f)?;
            for d_col in 0..3 {
                write(f, self.sticker_at(Face::D, Face::F, d_col, d_row))?;
            }
            f.write_char('\n')?;
        }

        Ok(())
    }
}

#[cfg(all(test, feature = "quickcheck"))]
mod tests {
    use quickcheck::{Arbitrary, Gen};

    use super::*;

    impl Arbitrary for Cube {
        fn arbitrary(g: &mut Gen) -> Self {
            let scramble = <Vec<Move>>::arbitrary(g);
            Cube::SOLVED.mov(scramble)
        }
    }

    #[test]
    fn debug_cube_insta() {
        insta::assert_debug_snapshot!(Cube::SOLVED)
    }

    #[test]
    fn display_cube_insta() {
        insta::assert_snapshot!(Cube::SOLVED)
    }
}
