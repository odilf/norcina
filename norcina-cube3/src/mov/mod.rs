pub mod alg;

pub use alg::algs;

#[cfg(all(test, feature = "quickcheck"))]
mod tests {
    pub use super::*;
    use crate::cube::Cube;
    use norcina_core::mov::InvertibleMove;
    use norcina_cube_n::{
        alg,
        math::Face,
        mov::{Amount, Move},
    };
    use quickcheck::quickcheck;

    quickcheck! {
        fn fn_move_constructor_and_accessors_maintain_values(face: Face, amount: Amount) -> bool {
            let mov = Move::new(face, amount);
            mov.face() == face && mov.amount() == amount
        }

        fn fn_double_double_identity(cube: Cube, face: Face) -> bool {
            let mov = Move::new(face, Amount::Double);
            cube.mov([mov, mov]) == cube
        }

        fn fn_rrp_identity(cube: Cube) -> bool {
            cube.mov(alg!(R RP)) == cube
        }

        fn fn_move_reverse_identity(cube: Cube, face: Face) -> bool {
            let m1 = Move::new(face, Amount::Single);
            let m2 = Move::new(face, Amount::Reverse);
            cube.mov([m1, m2]) == cube
        }

        fn fn_double_t_identity(cube: Cube) -> bool {
            cube.mov(alg!(R RP)) == cube
        }

        fn fn_single_double_equals_reverse(cube: Cube) -> bool {
            cube.mov(alg!(R R2)) == cube.mov(alg!(RP))
        }

        fn move_and_reverse_identity(cube: Cube, mov: Move) -> bool {
            cube.mov_single(mov).mov_single(mov.inverse()) == cube
        }
    }

    #[test]
    fn instance_all_basic_moves() {
        for mov in Move::iter() {
            println!("Move is {mov}");
            insta::assert_debug_snapshot!(Cube::SOLVED.mov_single(mov))
        }
    }

    #[test]
    fn ua_ub_cancel() {
        assert!(
            Cube::SOLVED
                .mov(algs::pll::U_A)
                .mov(algs::pll::U_B)
                .is_solved()
        )
    }
}
