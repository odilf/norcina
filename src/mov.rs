use std::ops;

use crate::math::{Axis, Direction, Face};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Amount {
    Single = 1,
    Double = 2,
    Reverse = 3,
}

impl Amount {
    #[inline]
    pub const fn u8(self) -> u8 {
        self as u8
    }

    #[inline]
    pub const fn from_u8(x: u8) -> Self {
        match x {
            1 => Self::Single,
            2 => Self::Double,
            3 => Self::Reverse,
            _ => panic!("Invalid amount."),
        }
    }
}

impl ops::Mul<Direction> for Amount {
    type Output = Amount;
    fn mul(self, rhs: Direction) -> Self::Output {
        // TODO: Is it possible to remove this branch?
        match rhs {
            Direction::Positive => self,
            Direction::Negative => Self::from_u8(4 - self.u8()),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Move {
    /// Packed field: `---aafff`
    data: u8,
}

impl Move {
    pub const fn new(face: Face, amount: Amount) -> Move {
        Self {
            data: face.u8() + (amount.u8() << 3),
        }
    }

    pub const fn face(self) -> Face {
        Face::from_u8(self.data & 0b111)
    }

    pub const fn amount(self) -> Amount {
        Amount::from_u8(self.data >> 3)
    }

    pub const fn axis(self) -> Axis {
        self.face().axis()
    }
}

#[macro_export]
macro_rules! alg {
    (@ $mov:tt) => { $mov };

    ($($mov:tt)*) => {{
        use $crate::mov::moves::*;
        [$(alg!(@ $mov)),*]
    }};

}

pub mod moves {
    macro_rules! generate_moves {
        // ($face:ident, $amount:ident, $amount_name:tt) => {
        //     pub const $face: Move = Move::new($face, $amount);
        // };

        // ($face:ident, [$($amounts:ident)*], [$($amount_names:tt)*]) => {
        //     $(
        //         generate_moves!($face, $amounts, $amount_names);
        //     )*
        // };
        (@amount 1) => { Amount::Single };
        (@amount 2) => { Amount::Double };
        (@amount 3) => { Amount::Reverse };

        (@face R) => { Face::R };
        (@face U) => { Face::U };
        (@face F) => { Face::F };
        (@face L) => { Face::L };
        (@face D) => { Face::D };
        (@face B) => { Face::B };

        ([$($face:tt $amount:tt $name:ident),*]) => {
            $(
                pub const $name: Move = Move::new(
                    generate_moves!(@face $face),
                    generate_moves!(@amount $amount)
                );
            )*
        };
    }

    use super::Amount;
    use super::Face;
    use super::Move;

    generate_moves!([
        R 1 R, R 2 R2, R 3 RP,
        U 1 U, U 2 U2, U 3 UP,
        F 1 F, F 2 F2, F 3 FP,
        L 1 L, L 2 L2, L 3 LP,
        D 1 D, D 2 D2, D 3 DP,
        B 1 B, B 2 B2, B 3 BP
    ]);
}

pub mod algs {
    use super::Move;

    pub const SEXY: [Move; 4] = alg!(R U RP UP);
    pub const SLEDGEHAMMER: [Move; 4] = alg!(RP F R FP);
    pub const T: [Move; 14] = alg!(R U RP UP RP F R2 UP RP UP R U RP FP);
    pub const J: [Move; 13] = alg!(R U RP F R U RP UP RP FP R2 UP RP);

    pub const CHECKER: [Move; 6] = alg!(R2 L2 U2 D2 F2 B2);

    // TODO: Concat or extend algs
    // pub const J_AUF: [Move; 14] = [J, alg!(UP)].concat();
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::cube::Cube;
    use quickcheck::{Arbitrary, Gen, quickcheck};

    impl Arbitrary for Amount {
        fn arbitrary(g: &mut Gen) -> Self {
            *g.choose(&[Amount::Single, Amount::Double, Amount::Reverse])
                .unwrap()
        }
    }

    impl Arbitrary for Move {
        fn arbitrary(g: &mut Gen) -> Self {
            let face = Face::arbitrary(g);
            let amount = Amount::arbitrary(g);
            Move::new(face, amount)
        }
    }

    quickcheck! {
        fn fn_move_constructor_and_accessors_maintain_values(face: Face, amount: Amount) -> bool {
            let mov = Move::new(face, amount);
            mov.face() == face && mov.amount() == amount
        }

        fn fn_double_double_identity(cube: Cube, face: Face) -> bool {
            let mov = Move::new(face, Amount::Double);
            cube.mov([mov, mov]) == cube
        }

        fn fn_rrp_identity(cube: Cube) -> bool {
            cube.mov(alg!(R RP)) == cube
        }

        fn fn_move_reverse_identity(cube: Cube, face: Face) -> bool {
            let m1 = Move::new(face, Amount::Single);
            let m2 = Move::new(face, Amount::Reverse);
            cube.mov([m1, m2]) == cube
        }

        fn fn_double_t_identity(cube: Cube) -> bool {
            cube.mov(alg!(R RP)) == cube
        }

        fn fn_single_double_equals_reverse(cube: Cube) -> bool {
            cube.mov(alg!(R R2)) == cube.mov(alg!(RP))
        }
    }
}
