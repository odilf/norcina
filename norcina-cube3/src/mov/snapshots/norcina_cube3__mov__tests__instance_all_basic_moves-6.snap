---
source: norcina-cube3/src/mov/mod.rs
expression: "Cube::SOLVED.mov_single(mov)"
---
Cube {
    corners: [
        LUF (2) is at RUF,
        LUB (1) is at LUF,
        RDF (0) is at RDF,
        LDF (0) is at LDF,
        RUF (1) is at RUB,
        RUB (2) is at LUB,
        RDB (0) is at RDB,
        LDB (0) is at LDB,
    ],
    edges: [
        LU (✓) is at UF,
        DF (✓) is at DF,
        RU (✓) is at UB,
        DB (✓) is at DB,
        FR (✓) is at FR,
        BR (✓) is at BR,
        FL (✓) is at FL,
        BL (✓) is at BL,
        UF (✓) is at RU,
        UB (✓) is at LU,
        RD (✓) is at RD,
        LD (✓) is at LD,
    ],
}
